# BEE 4750 Homework 4: Linear Programming and Capacity Expansion

**Name**: Sarah Cooke, Fabien De Silva, Laura Akey

**ID**: sec275, fmd48, lka27

> **Due Date**
>
> Thursday, 11/07/23, 9:00pm

## Overview

### Instructions

-   Problem 1 asks you to formulate and solve a resource allocation
    problem using linear programming.
-   Problem 2 asks you to formulate, solve, and analyze a standard
    generating capacity expansion problem.
-   Problem 3 (5750 only) asks you to add a CO<sub>2</sub> constraint to
    the capacity expansion problem and identify changes in the resulting
    solution.

### Load Environment

The following code loads the environment and makes sure all needed
packages are installed. This should be at the start of most Julia
scripts.

In [1]:
import Pkg
Pkg.activate(@__DIR__)
Pkg.instantiate()

  Activating project at `~/hw04`


## Problems (Total: 50/60 Points)

### Problem 1 (20 points)

A farmer has access to a pesticide which can be used on corn, soybeans,
and wheat fields, and costs \$70/kg-yr to apply. The crop yields the
farmer can obtain following crop yields by applying varying rates of
pesticides to the field are shown in
<a href="#tbl-yields" class="quarto-xref">Table 1</a>.

| Application Rate (kg/ha) | Soybean (kg/ha) | Wheat (kg/ha) | Corn (kg/ha) |
|:------------------------:|:---------------:|:-------------:|:------------:|
|            0             |      2900       |     3500      |     5900     |
|            1             |      3800       |     4100      |     6700     |
|            2             |      4400       |     4200      |     7900     |

Table 1: Crop yields from applying varying pesticide rates for Problem
1.

The costs of production, *excluding pesticides*, for each crop, and
selling prices, are shown in
<a href="#tbl-costs" class="quarto-xref">Table 2</a>.

|   Crop   | Production Cost (\$/ha-yr) | Selling Price (\$/kg) |
|:--------:|:--------------------------:|:---------------------:|
| Soybeans |            350             |         0.36          |
|  Wheat   |            280             |         0.27          |
|   Corn   |            390             |         0.22          |

Table 2: Costs of crop production, excluding pesticides, and selling
prices for each crop.

Recently, environmental authorities have declared that farms cannot have
an *average* application rate on soybeans, wheat, and corn which exceeds
0.8, 0.7, and 0.6 kg/ha, respectively. The farmer has asked you for
advice on how they should plant crops and apply pesticides to maximize
profits over 130 total ha while remaining in regulatory compliance if
demand for each crop (which is the maximum the market would buy) this
year is 250,000 kg?

**In this problem**:

-   Formulate a linear program for this resource allocation problem,
    including clear definitions of decision variable(s) (including
    units), objective function(s), and constraint(s) (make sure to
    explain functions and constraints with any needed derivations and
    explanations). **Tip: Make sure that all of your constraints are
    linear**.
-   Implement the program in `JuMP.jl` and find the solution. How many
    ha should the farmer dedicate to each crop and with what pesticide
    application rate(s)? How much profit will the farmer expect to make?
-   The farmer has an opportunity to buy an extra 10 ha of land. How
    much extra profit would this land be worth to the farmer? Discuss
    why this value makes sense and whether you would recommend the
    farmer should make the purchase.

In [2]:
using JuMP
using HiGHS
using DataFrames
using Plots
using Measures
using CSV
using MarkdownTables

$Problem$ $1$ $Solution$


note: S[0:2], W[0:2], and C[0:2] are in units ha

variables are defined as the values from the table above

**Soybeans:**

soybean (kg) = soybean(0) (kg/ha) *S[0] (ha) + soybean(1) (kg/ha) *S[1] (ha) + soybean(2) (kg/ha) *S[2] (ha)

sales_soybeans (dollar) = 0.36 ($/kg) * soybean(kg)

cost_pesticides_soybean (dollar/yr) = 70 ($/kg-yr) * (0*S[0] + 1*S[1] + 2*S[2]) (kg)

cost_production_soybean (dollar/yr) = 350 ($/ha-yr) * (S[0] + S[1] + S[2]) (ha)


profit_soybean ($)= sales_soybean - cost_pesticides_soybean - cost_production_soybean

exceedance (kg/ha) * total_land_soybean (ha) ≥ total_pesticide_amount (kg)

numberical equation values:

**profit_soybean = 694*S[0] + 948*S[1] + 1094*S[2]**

**0.8 * (S[0] + S[1] + S[2]) ≥ (0*S[0] + 1*S[1] + 2*S[2])**


**Wheat:**

wheat (kg) = wheat(0) (kg/ha) *W[0] (ha) + wheat(1) (kg/ha) *W[1] (ha) + wheat(2) (kg/ha) *W[2] (ha)

sales_wheat (dollar) = 0.27 ($/kg) * wheat(kg)

cost_pesticides_wheat (dollar/yr) = 70 ($/kg-yr) * (0*W[0] + 1*W[1] + 2*W[2]) (kg)

cost_production_wheat (dollar/yr) = 280 ($/ha-yr) * (W[0] + W[1] + W[2]) (ha)


profit_wheat ($)= sales_wheat - cost_pesticides_wheat - cost_production_wheat

exceedance (kg/ha) * total_land_wheat (ha) ≥ total_pesticide_amount (kg)

numberical equation values:


**profit_wheat = 665*W[0] + 757*W[1] + 714*W[2]**

**0.7 * (W[0] + W[1] + W[2]) ≥ (0*W[0] + 1*W[1] + 2*W[2])**


**Corn:**

corn (kg) = corn(0) (kg/ha) *C[0] (ha) + corn(1) (kg/ha) *C[1] (ha) + corn(2) (kg/ha) *C[2] (ha)

sales_corn (dollar) = 0.22 ($/kg) * corn(kg)

cost_pesticides_corn (dollar/yr) = 70 ($/kg-yr) * (0*C[0] + 1*C[1] + 2*C[2]) (kg)

cost_production_corn (dollar/yr) = 390 ($/ha-yr) * (C[0] + C[1] + C[2]) (ha)


profit_corn ($)= sales_corn - cost_pesticides_corn - cost_production_corn

exceedance (kg/ha) * total_land_corn (ha) ≥ total_pesticide_amount (kg)

numberical equation values:


**profit_corn = 908*C[0] + 1014*C[1] + 1208*C[2]**

**0.6 * (C[0] + C[1] + C[2]) ≥ (0*C[0] + 1*C[1] + 2*C[2])**


**Overall:**

130 ≥ S[0] + S[1] + S[2] + W[0] + W[1] + W[2] + C[0] + C[1] + C[2]

250,000 ≥ soybean (kg) 

250,000 ≥ wheat (kg)

250,000 ≥ corn (kg)


**Constraints:**

pesticides cost 70 $/kg-yr

average application rate of soybean, wheat, and corn must be below 0.8, 0.7, and 0.6 kg/ha respectively.

soybean ha + wheat ha + corn ha ≤ 130 ha

deamnd for each crop ≤ 250,000 kg/crop

In [32]:
#soybeans
S_prod_0 = 2900
S_prod_1 = 3800
S_prod_2 = 4400
S_sellPrice = 0.36
S_prod_cost = 350

#wheat
W_prod_0 = 3500
W_prod_1 = 4100
W_prod_2 = 4200
W_sellPrice = 0.27
W_prod_cost = 280

#corn
C_prod_0 = 5900
C_prod_1 = 6700
C_prod_2 = 7900
C_sellPrice = 0.22
C_prod_cost = 390

#overall
price_pesticides = 70
demand_max = 250000

#start model
crops_model = Model(HiGHS.Optimizer) # initialize model object
@variable(crops_model, S[0:2] >= 0) # non-negativity constraints
@variable(crops_model, W[0:2] >= 0) # non-negativity constraints
@variable(crops_model, C[0:2] >= 0) # non-negativity constraints

#soybeans
soy_kg = S[0]*S_prod_0 + S[1]* S_prod_1 + S[2] * S_prod_2
soy_sell = S_sellPrice * soy_kg
soy_pesticides_cost = price_pesticides * (S[1] + 2*S[2])
soy_prod_cost = S_prod_cost * sum(S)

soy_profit = soy_sell - soy_pesticides_cost - soy_prod_cost

#wheat
wheat_kg = W[0]*W_prod_0 + W[1]* W_prod_1 + W[2] * W_prod_2
wheat_sell = W_sellPrice * wheat_kg
wheat_pesticides_cost = price_pesticides * (W[1] + 2*W[2])
wheat_prod_cost = W_prod_cost * sum(W)

wheat_profit = wheat_sell - wheat_pesticides_cost - wheat_prod_cost

#corn
corn_kg = C[0]*C_prod_0 + C[1]* C_prod_1 + C[2] * C_prod_2
corn_sell = C_sellPrice * corn_kg
corn_pesticides_cost = price_pesticides * (C[1] + 2 * C[2])
corn_prod_cost = C_prod_cost * sum(C)

corn_profit = corn_sell - corn_pesticides_cost - corn_prod_cost

sell = soy_sell + wheat_sell + corn_sell
prod_cost = soy_prod_cost + wheat_prod_cost + corn_prod_cost
pest_cost = soy_pesticides_cost + wheat_pesticides_cost + corn_pesticides_cost

profit = sell - prod_cost - pest_cost

@objective(crops_model, Max, profit)
@constraint(crops_model, soybean, S[1] + 2*S[2] <= 0.8 * sum(S))
@constraint(crops_model, wheat, W[1]+2*W[2] <= 0.7 * (sum(W)))
@constraint(crops_model, corn, C[1]+2*C[2] <= 0.6 * (sum(W)))
@constraint(crops_model, soybean_demand, soy_kg <= demand_max)
@constraint(crops_model, wheat_demand, wheat_kg <= demand_max)
@constraint(crops_model, corn_demand, corn_kg <= demand_max)
@constraint(crops_model, land_constraint, sum(S) + sum(W) + sum(C) <= 130)

optimize!(crops_model)

profit = objective_value(crops_model)
println("Total profit: ", profit)

Running HiGHS 1.7.2 (git hash: 5ce7a2753): Copyright (c) 2024 HiGHS under MIT licence terms
Coefficient ranges:
  Matrix [2e-01, 8e+03]
  Cost   [7e+02, 1e+03]
  Bound  [0e+00, 0e+00]
  RHS    [1e+02, 2e+05]
Presolving model
7 rows, 9 cols, 29 nonzeros  0s
7 rows, 9 cols, 29 nonzeros  0s
Presolve : Reductions: rows 7(-0); columns 9(-0); elements 29(-0) - Not reduced
Problem not reduced by presolve: solving the LP
Using EKK dual simplex solver - serial
  Iteration        Objective     Infeasibilities num(sum)
          0    -1.6206236852e+02 Ph1: 6(22.6083); Du: 9(162.062) 0s
          7     1.1546294152e+05 Pr: 0(0) 0s
Model   status      : Optimal
Simplex   iterations: 7
Objective value     :  1.1546294152e+05
HiGHS run time      :          0.00
Total profit: 115462.94151985823


In [25]:
@show value.(S);

value.(S) = 1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
 13.812154696132595
 55.24861878453037
  0.0


In [26]:
@show value.(W);

value.(W) = 1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
  6.200458667778595
 14.467736891483387
  0.0


In [27]:
@show value.(C);

value.(C) = 1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
 34.07057229229646
  0.0
  6.200458667778598


In [28]:
@show shadow_price(land_constraint)

shadow_price(land_constraint) = 809.3698113207549


809.3698113207549

**Solution:**

**A)** written above

**B)**

soybeans: 13.8 ha at 0 kg/ha, 55.2 ha at 1 kg/ha, and 0 ha at 2 kg/ha.

wheat: 6.2 ha at 0 kg/ha, 14.5 ha at 1 kg/ha, and 0 ha at 2 kg/ha.

corn: 34.1 ha at 0 kg/ha, 0 ha at 1 kg/ha, and 6.2 ha at 2 kg/ha.

This would give a profit of $115,462.94.

**C)**

In [33]:
#soybeans
S_prod_0 = 2900
S_prod_1 = 3800
S_prod_2 = 4400
S_sellPrice = 0.36
S_prod_cost = 350

#wheat
W_prod_0 = 3500
W_prod_1 = 4100
W_prod_2 = 4200
W_sellPrice = 0.27
W_prod_cost = 280

#corn
C_prod_0 = 5900
C_prod_1 = 6700
C_prod_2 = 7900
C_sellPrice = 0.22
C_prod_cost = 390

#overall
price_pesticides = 70
demand_max = 250000

#start model
crops_model = Model(HiGHS.Optimizer) # initialize model object
@variable(crops_model, S[0:2] >= 0) # non-negativity constraints
@variable(crops_model, W[0:2] >= 0) # non-negativity constraints
@variable(crops_model, C[0:2] >= 0) # non-negativity constraints

#soybeans
soy_kg = S[0]*S_prod_0 + S[1]* S_prod_1 + S[2] * S_prod_2
soy_sell = S_sellPrice * soy_kg
soy_pesticides_cost = price_pesticides * (S[1] + 2*S[2])
soy_prod_cost = S_prod_cost * sum(S)

soy_profit = soy_sell - soy_pesticides_cost - soy_prod_cost

#wheat
wheat_kg = W[0]*W_prod_0 + W[1]* W_prod_1 + W[2] * W_prod_2
wheat_sell = W_sellPrice * wheat_kg
wheat_pesticides_cost = price_pesticides * (W[1] + 2*W[2])
wheat_prod_cost = W_prod_cost * sum(W)

wheat_profit = wheat_sell - wheat_pesticides_cost - wheat_prod_cost

#corn
corn_kg = C[0]*C_prod_0 + C[1]* C_prod_1 + C[2] * C_prod_2
corn_sell = C_sellPrice * corn_kg
corn_pesticides_cost = price_pesticides * (C[1] + 2 * C[2])
corn_prod_cost = C_prod_cost * sum(C)

corn_profit = corn_sell - corn_pesticides_cost - corn_prod_cost

sell = soy_sell + wheat_sell + corn_sell
prod_cost = soy_prod_cost + wheat_prod_cost + corn_prod_cost
pest_cost = soy_pesticides_cost + wheat_pesticides_cost + corn_pesticides_cost

profit = sell - prod_cost - pest_cost

@objective(crops_model, Max, profit)
@constraint(crops_model, soybean, S[1] + 2*S[2] <= 0.8 * sum(S))
@constraint(crops_model, wheat, W[1]+2*W[2] <= 0.7 * (sum(W)))
@constraint(crops_model, corn, C[1]+2*C[2] <= 0.6 * (sum(W)))
@constraint(crops_model, soybean_demand, soy_kg <= demand_max)
@constraint(crops_model, wheat_demand, wheat_kg <= demand_max)
@constraint(crops_model, corn_demand, corn_kg <= demand_max)
@constraint(crops_model, land_constraint, sum(S) + sum(W) + sum(C) <= 140)

optimize!(crops_model)

profit = objective_value(crops_model)
println("Total profit: ", profit)

Running HiGHS 1.7.2 (git hash: 5ce7a2753): Copyright (c) 2024 HiGHS under MIT licence terms
Coefficient ranges:
  Matrix [2e-01, 8e+03]
  Cost   [7e+02, 1e+03]
  Bound  [0e+00, 0e+00]
  RHS    [1e+02, 2e+05]
Presolving model
7 rows, 9 cols, 29 nonzeros  0s
7 rows, 9 cols, 29 nonzeros  0s
Presolve : Reductions: rows 7(-0); columns 9(-0); elements 29(-0) - Not reduced
Problem not reduced by presolve: solving the LP
Using EKK dual simplex solver - serial
  Iteration        Objective     Infeasibilities num(sum)
          0    -1.6206236852e+02 Ph1: 6(22.6083); Du: 9(162.062) 0s
          7     1.2355663963e+05 Pr: 0(0) 0s
Model   status      : Optimal
Simplex   iterations: 7
Objective value     :  1.2355663963e+05
HiGHS run time      :          0.00
Total profit: 123556.63963306576


Solutions(continued):

**C)**

With an additional 10 ha, the new profit is $123,556.64 which means that it makes $8,093.70 more each year. Our recommendation depends on the price of purchasing the land and noneconomic considerations from the additional labor. If this difference is trivial and the purchasing price is such that the profit covers it in a few years, then this is a worthy purchase.

### Problem 2 (30 points)

For this problem, we will use hourly load (demand) data from 2013 in New
York’s Zone C (which includes Ithaca). The load data is loaded and
plotted below in <a href="#fig-demand" class="quarto-xref">Figure 1</a>.

In [1]:
# load the data, pull Zone C, and reformat the DataFrame
NY_demand = DataFrame(CSV.File("data/2013_hourly_load_NY.csv"))
rename!(NY_demand, :"Time Stamp" => :Date)
demand = NY_demand[:, [:Date, :C]]
rename!(demand, :C => :Demand)
demand[:, :Hour] = 1:nrow(demand)

# plot demand
plot(demand.Hour, demand.Demand, xlabel="Hour of Year", ylabel="Demand (MWh)", label=:false)

Next, we load the generator data, shown in
<a href="#tbl-generators" class="quarto-xref">Table 3</a>. This data
includes fixed costs (\$/MW installed), variable costs (\$/MWh
generated), and CO<sub>2</sub> emissions intensity (tCO<sub>2</sub>/MWh
generated).

In [1]:
gens = DataFrame(CSV.File("data/generators.csv"))

Finally, we load the hourly solar and wind capacity factors, which are
plotted in <a href="#fig-cf" class="quarto-xref">Figure 2</a>. These
tell us the fraction of installed capacity which is expected to be
available in a given hour for generation (typically based on the average
meteorology).

In [1]:
# load capacify factors into a DataFrame
cap_factor = DataFrame(CSV.File("data/wind_solar_capacity_factors.csv"))

# plot January capacity factors
p1 = plot(cap_factor.Wind[1:(24*31)], label="Wind")
plot!(cap_factor.Solar[1:(24*31)], label="Solar")
xaxis!("Hour of the Month")
yaxis!("Capacity Factor")

p2 = plot(cap_factor.Wind[4344:4344+(24*31)], label="Wind")
plot!(cap_factor.Solar[4344:4344+(24*31)], label="Solar")
xaxis!("Hour of the Month")
yaxis!("Capacity Factor")

display(p1)
display(p2)

You have been asked to develop a generating capacity expansion plan for
the utility in Riley County, NY, which currently has no existing
electrical generation infrastructure. The utility can build any of the
following plant types: geothermal, coal, natural gas combined cycle gas
turbine (CCGT), natural gas combustion turbine (CT), solar, and wind.

While coal, CCGT, and CT plants can generate at their full installed
capacity, geothermal plants operate at maximum 85% capacity, and solar
and wind available capacities vary by the hour depend on the expected
meteorology. The utility will also penalize any non-served demand at a
rate of \$10,000/MWh.

**In this problem**:

-   Formulate a linear program for this capacity expansion problem,
    including clear definitions of decision variable(s) (including
    units), objective function(s), and constraint(s) (make sure to
    explain functions and constraints with any needed derivations and
    explanations).
-   Implement your linear program in `JuMP.jl`. Find the optimal
    solution. How much should the utility build of each type of
    generating plant? What will the total cost be? How much energy will
    be non-served?
-   What fraction of annual generation does each plant type produce? How
    does this compare to the breakdown of built capacity that you found
    in Problem 1.5? Do these results make sense given the generator
    data?
-   Make a plot of the electricity price in each hour. Discuss any
    trends that you see.

> **Significant Digits**
>
> Use `round(x; digits=n)` to report values to the appropriate
> precision! If your number is on a different order of magnitude and you
> want to round to a certain number of significant digits, you can use
> `round(x; sigdigits=n)`.

> **Getting Variable Output Values**
>
> `value.(x)` will report the values of a `JuMP` variable `x`, but it
> will return a special container which holds other information about
> `x` that is useful for `JuMP`. This means that you can’t use this
> output directly for further calculations. To just extract the values,
> use `value.(x).data`.

> **Suppressing Model Command Output**
>
> The output of specifying model components (variable or constraints)
> can be quite large for this problem because of the number of time
> periods. If you end a cell with an `@variable` or `@constraint`
> command, I *highly* recommend suppressing output by adding a
> semi-colon after the last command, or you might find that your
> notebook crashes.

### Problem 3 (10 points)

**This problem is only required for students in BEE 5750**.

The NY state legislature is considering enacting an annual
CO<sub>2</sub> limit, which for the utility would limit the emissions in
its footprint to 1.5 MtCO<sub>2</sub>/yr.

**In this problem**:

-   Reformulate your linear program from Problem 2 with any necessary
    changes to capture the CO<sub>2</sub> limit.
-   Implement the new optimization problem and find the optimal
    solution. How much should the utility build of each type of
    generating plant? What is different from your plan from Problem 1?
    Do these changes make sense?
-   What would the value to the utility be of allowing it to emit an
    additional 1000 tCO<sub>2</sub>/yr? An additional 5000?

## References

List any external references consulted, including classmates.